Add expression tabulation test

michalhabera opened the pull request:
Based on https://github.com/FEniCS/ffcx/pull/160

The test shows an interpolation of grad operator seen as mapping `P2` to vector `dP1`.

michalhabera commented on python/test/unit/fem/test_interpolation.py:
```diff
@@ -0,0 +1,41 @@
+
+    dolfin.fem.assemble_vector(f.vector, L, mode=dolfin.cpp.fem.InsertMode.set)
+
+    # Update ghosts with the same mode as assembly
```
Hmm, what do you mean with the mode "update ghosts"? The standard assembler with "add" mode needs update ghosts with add mode, "insert" assembler needs insert ghost update.

Ghost update with add wouldn't work in this case, as for interpolation local information must be forced to the global vector, not added. I am missing something?

IgorBaratta replied:
`ghostUpdate(addv=PETSc.InsertMode.INSERT, mode=PETSc.ScatterMode.FORWARD)`  means  that you're updating the ghost values with the values from the owning processes. It's just the standard mode to update the values of the ghosts. And It would not make sense to use add in that case. 

Maybe you're referring to  `ScatterMode.REVERSE`, in the sense of applying the ghost values to the owning process.

michalhabera replied:
Ah, correct, I should be scattering with REVERSE mode. Insert local ghost information back into the global vector. Thats what you mean?

chrisrichardson replied:
I'm also not clear why you even need to update ghosts for `dP0` space?

michalhabera replied:
Not needed to communicate... :) I'll remove.

IgorBaratta replied:
And also, applying reverse scatter mode with `INSERT`, in a ghosted mesh, would make the tests to fail.

garth-wells commented on python/setup.py:
```diff
@@ -67,7 +67,7 @@ def build_extension(self, ext):
                 build_args += ['--', '-j3']
             else:
-                num_build_threads = max(1, multiprocessing.cpu_count() - 1)
+                num_build_threads = min(3, max(1, multiprocessing.cpu_count() - 1))
```
What's up here?

garth-wells commented on python/src/fem.cpp:
```diff
@@ -213,7 +213,8 @@ void fem(py::module& m)
                     dolfin::fem::DirichletBC::Method>(),
            py::arg("V"), py::arg("g"), py::arg("facets"), py::arg("method"))
-      .def("function_space", &dolfin::fem::DirichletBC::function_space);
+      .def("function_space", &dolfin::fem::DirichletBC::function_space)
```
Should this be a property?

garth-wells commented on python/test/unit/fem/test_interpolation.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def test_interpolation_laplace():
+    mesh = dolfin.generation.UnitSquareMesh(dolfin.MPI.comm_world, 5, 5)
```
Add a brief docstring to explain what the test does.

garth-wells commented on python/test/unit/fem/test_interpolation.py:
```diff
@@ -0,0 +1,37 @@
+    ufl_expr = ufl.div(ufl.grad(u))
+
+    # Attach the expression with cell midpoint
+    expr = (ufl_expr, np.array([[0.25, 0.25]]))
```
Explain what the midpoint is for.

michalhabera commented:
@garth-wells I removed unrelated changes. Also changed the test to compute `grad()` instead of laplace. Grad also picks issues in ordering, is less trivial.

michalhabera commented:
Assembler with set mode was removed.
